docs: add v0.5.0 blog post (#309)
posit-dev/great-tables: docs/blog/introduction-0.5.0/index.qmd

We're happy to share that Great Tables v0.5.0 makes Pandas an optional dependency.
Using Pandas DataFrames as inputs is still fully supported.
The optional dependencies simply allow users of one DataFrame library to not have to install the others.

In this post, I'll cover three important pieces:

* Where we are today on dependencies
* The challenge of removing hard dependencies
* How we made pandas optional

## The state of Great Tables dependencies

Currently, Great Tables has two "sizes" of libraries it depends on:

* **Small**: utility libraries for things like web colors, and datetime localization.
* **Big**: a lingering dependency on numpy in a few places (like nanoplots).

For small utilities, we depend on Babel, which makes it easier to say things like, "format this number as if I'm in Germany". We also depend on webcolors, which allows us to know that "goldenrod" is the hex color "#daa520".

For big dependencies, numpy should be fairly straightforward to remove (see [this issue](https://github.com/posit-dev/great-tables/issues/296)).

> We also still really on pandas for datasets in `great_tables.data`, but will remove it soon (see [this issue](https://github.com/posit-dev/great-tables/issues/91)).

Reviewer: really on => rely on ?

Author: Thanks for catching these typos!

> Removing dependencies like numpy and Pandas helps people who are in restricted computing environments, want a more lightweight install, or who are stuck depending on an much earlier version of a package.

Reviewer: an much earlier version of a package => a much earlier version of a package ?

---
title: "Great Tables: Pandas now an Optional Dependency"
html-table-processing: none
author: Michael Chow
date: 2024-04-18
freeze: true
jupyter: python3
---

A few months ago, we released a [blog post](./polars-styling) about how much we loved the combination of Polars and Great Tables. We found that Polars lazy expression system opened up convenient ways to conditionally format tables for presentation. However, excited as we were, we were harboring a shameful secret: Great Tables enabled Polars as an optional dependency, but had a hard dependency on the alternative DataFrame library Pandas.

We're happy to share that Great Tables v0.5.0 makes Pandas an optional dependency.
Using Pandas DataFrames as inputs is still fully supported.

> The optional dependency simply allow users of one DataFrame library to not have to install the others.

Reviewer 2: ```suggestion
The optional dependency simply allows users of one DataFrame library to not have to install the other.
```

In this post, I'll cover three important pieces:

* Where we are today on dependencies
* The challenge of removing hard dependencies

> * How we made pandas optional

Reviewer 2: ```suggestion
* How we made Pandas optional
```

## The state of Great Tables dependencies

Currently, Great Tables has two "sizes" of libraries it depends on:

> * **Small**: utility libraries for things like web colors, and datetime localization.

Reviewer 2: ```suggestion
* **Small**: utility libraries for things like datetime localization.
```

* **Big**: a lingering dependency on numpy in a few places (like nanoplots).

For small utilities, we depend on Babel, which makes it easier to say things like, "format this number as if I'm in Germany".

For big dependencies, numpy should be fairly straightforward to remove (see [this issue](https://github.com/posit-dev/great-tables/issues/296)).

> We also still rely on pandas for datasets in `great_tables.data`, but will remove it soon (see [this issue](https://github.com/posit-dev/great-tables/issues/91)).

Reviewer 2: ```suggestion
We also still rely on Pandas for datasets in `great_tables.data`, but we will remove it soon (see [this issue](https://github.com/posit-dev/great-tables/issues/91)).
```

Removing dependencies like numpy and Pandas helps people who are in restricted computing environments, want a more lightweight install, or who are stuck depending on a much earlier version of a package.
It also helps us keep a clean separation of concerns.  Without clear boundaries, it's too tempting to reach for things like `pd.isna()` in a pinch,
or smear library specific versions of missingness across our code (e.g. `pd.NA`, `np.nan`, `polars.NullType`).

## The challenge of removing hard dependencies

> Removing hard dependencies on DataFrame libraries is worthwhile, but requires special handling for all DataFrame specific actions. To illustrate consider the Great Tables output below, which is produced from a pandas DataFrame:

Reviewer 2: ```suggestion
Removing hard dependencies on DataFrame libraries is worthwhile, but requires special handling for all DataFrame specific actions. To illustrate consider the Great Tables output below, which is produced from a Pandas DataFrame:
```

```{python}
# get columns as a list ----

df_pandas.columns.tolist()
df_polars.columns
```

Notice that the two lines of code aren't too different---Pandas just requires an extra `.tolist()` piece.
We could create a special function, that returns a list of names, depending on the type of the input DataFrame.

```{python}
def get_column_names(data) -> list[str]:

    # pandas specific ----
    if isinstance(data, pd.DataFrame):
        return data.columns.tolist()

    # polars specific ----
    elif isinstance(data, pl.DataFrame):
        return data.columns

    raise TypeError(f"Unsupported type {type(data)}")
```

The function works great, in that we can call it on either DataFrame, but it lacks two things.
The first is **dependency inversion**, since it requires importing both pandas and polars (creating a hard dependency).
The second is **separation of concerns**, since Pandas and Polars code is mixed together. In this case adding more DataFrame implementations would create a hot stew of logic.

## How we made Pandas optional

We were able to make Pandas optional in a sane manner through two moves:

> * [databackend](https://github.com/machow/databackend): perform isinstance checks without importing anything.

Reviewer 2: ```suggestion
* [databackend](https://github.com/machow/databackend): perform `isinstance` checks without importing anything.
```

* singledispatch: split out functions like `get_column_names()` into DataFrame specific versions.

### Inverting dependency with databackend

```{python}
from great_tables._databackend import AbstractBackend

class PdDataFrame(AbstractBackend):
    _backends = [("pandas", "DataFrame")]

class PlDataFrame(AbstractBackend):
    _backends = [("polars", "DataFrame")]

if isinstance(df_pandas, PdDataFrame):
    print("I'm a pandas DataFrame!!!")
```

> Note that the PdDataFrame above is able to detect a pandas DataFrame without importing pandas, by taking advantage of a bit of logic called a counterfactual:

Reviewer 2: ```suggestion
Note that the `PdDataFrame` above is able to detect a Pandas DataFrame without importing Pandas, by taking advantage of a bit of logic called a counterfactual:
```

* assumption: if `df_pandas` is a Pandas DataFrame, then Pandas has been imported.
* counterfactual: if Pandas has not been imported, then `df_pandas` is not a Pandas DataFrame.

This lets it quickly rule out a potential Pandas object by checking whether Pandas has been imported.
Since this can be done by looking inside `sys.modules`, no imports are required.
For more on this approach, see the [databackend README](https://github.com/machow/databackend).

### Separating concerns with singledispatch

> While databackend removes dependencies, the use of singledispatch from the built-in functools module separates out the logic for handling Polars DataFrom from the logic for Pandas DataFrames.

Reviewer 2: ```suggestion
While databackend removes dependencies, the use of singledispatch from the built-in `functools` module separates out the logic for handling Polars Dataframes from the logic for Pandas DataFrames.
```
